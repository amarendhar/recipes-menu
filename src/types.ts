export type Maybe<T> = T | null;

export type Recipe = {
  id: string;
  title: string;
  image: string;
  tags: string[];
  description: string;
  chefName: string;
};
