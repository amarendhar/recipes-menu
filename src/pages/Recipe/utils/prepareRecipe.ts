import { EntryFields, EntrySkeletonType } from "contentful";
import { Maybe, RecipeEntry, RecipeData } from "types";

type RecipeFields = {
  title: EntryFields.Text;
  photo: EntryFields.AssetLink;
  tags: EntryFields.Array<EntryFields.EntryLink<EntrySkeletonType>>;
  description: EntryFields.Text;
  chef: EntryFields.EntryLink<EntrySkeletonType>;
};

export const prepareRecipe = (data: Maybe<RecipeEntry>): Maybe<RecipeData> => {
  if (!data) {
    return null;
  }

  const fields = data?.fields as RecipeFields;

  const id: string = data.sys.id;
  const title = fields?.title || "";
  const image = (fields?.photo?.fields?.file?.url || "") as string;
  const tags =
    fields?.tags?.map((tag) => (tag?.fields?.name as string) || "") || [];
  const description = fields?.description || "";
  const chef = (fields?.chef?.fields?.name || "") as string;
  const rating = 0;

  const recipeData: RecipeData = {
    id,
    title,
    image,
    tags,
    description,
    chef,
    rating,
  };

  return recipeData;
};
