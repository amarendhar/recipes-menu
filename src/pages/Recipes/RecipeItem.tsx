import { useCallback } from "react";
import { NavLink } from "react-router-dom";
import styled, { keyframes } from "styled-components";
import { Rating } from "components";
import { UseRecipesProps } from "./useRecipes";
import { RecipesItemData } from "types";

type RecipeItemProps = {
  recipe: RecipesItemData;
  handleAddRating: UseRecipesProps["handleAddRating"];
};

export const RecipeItem = ({ recipe, handleAddRating }: RecipeItemProps) => {
  const onChangeRating = useCallback(
    (rating: number) => {
      handleAddRating({ recipeId: recipe.id, rating });
    },
    [handleAddRating, recipe.id]
  );

  return (
    <NavLink data-testid="recipe-item" to={`/recipe/${recipe.id}`}>
      <RecipeItemContainer>
        <RecipeImage>
          <img
            data-testid="recipe-image"
            src={recipe.image}
            alt={recipe.title}
          />
        </RecipeImage>
        <RecipeDetails>
          <RecipeName data-testid="recipe-name">{recipe.title}</RecipeName>
          <Rating value={recipe.rating} onChange={onChangeRating} />
        </RecipeDetails>
      </RecipeItemContainer>
    </NavLink>
  );
};

const fadeIn = keyframes`
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
`;

const RecipeItemContainer = styled.div`
  border-radius: ${({ theme }) => theme.shape.borderRadius};
  background-color: ${({ theme }) => theme.palette.primary.main};
  box-shadow: ${({ theme }) => theme.shadows[2]};
  animation: ${fadeIn} 0.2s ease-in;
  transition: all 0.2s ${({ theme }) => theme.transitions.easing.easeInOut} 0ms;
  overflow: hidden;

  &:hover {
    background-color: ${({ theme }) => theme.palette.primary.light};
    box-shadow: ${({ theme }) => theme.shadows[4]};
  }
`;

const RecipeImage = styled.div`
  position: relative;
  padding-top: 66.66%;
  overflow: hidden;

  img {
    position: absolute;
    width: 100%;
    inset: 0;
    overflow: hidden;
    transition: transform 0.2s
      ${({ theme }) => theme.transitions.easing.easeInOut} 0ms;
    background-color: ${({ theme }) => theme.palette.grey[500]};
  }
`;

const RecipeDetails = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  min-height: 100px;
  padding: ${({ theme }) => theme.spacing(3)};
`;

const RecipeName = styled.div`
  font-weight: ${({ theme }) => theme.typography.fontWeightBold};
`;
