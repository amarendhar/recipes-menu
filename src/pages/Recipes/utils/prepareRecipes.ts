import { Entry, EntryFields, EntrySkeletonType } from "contentful";
import { Maybe, RecipesEntry, RecipesItemData } from "types";

type RecipeFields = {
  title: EntryFields.Text;
  photo: EntryFields.AssetLink;
  tags: EntryFields.Array<EntryFields.EntryLink<EntrySkeletonType>>;
  description: EntryFields.Text;
  chef: EntryFields.EntryLink<EntrySkeletonType>;
};

export const prepareRecipes = (
  data: Maybe<RecipesEntry>
): RecipesItemData[] => {
  if (!data?.items) {
    return [];
  }

  const recipes: RecipesItemData[] = data.items.map(
    (recipe: Entry<EntrySkeletonType, undefined, string>) => {
      const fields = recipe?.fields as RecipeFields;

      const id: string = recipe.sys.id;
      const title = fields?.title || "";
      const image = (fields?.photo?.fields?.file?.url || "") as string;
      const rating = 0;

      const recipeData: RecipesItemData = {
        id,
        title,
        image,
        rating,
      };

      return recipeData;
    }
  );

  return recipes;
};
