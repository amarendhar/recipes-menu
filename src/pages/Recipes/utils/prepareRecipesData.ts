import {
  Entry,
  EntryFields,
  EntryCollection,
  EntrySkeletonType,
} from "contentful";
import { Maybe, Recipe } from "types";

type RecipeFields = {
  title: EntryFields.Text;
  photo: EntryFields.AssetLink;
  tags: EntryFields.Array<EntryFields.EntryLink<EntrySkeletonType>>;
  description: EntryFields.Text;
  chef: EntryFields.EntryLink<EntrySkeletonType>;
};

export const prepareRecipes = (
  data: Maybe<EntryCollection<EntrySkeletonType, undefined, string>>
): Recipe[] => {
  if (!data?.items) {
    return [];
  }

  const recipes: Recipe[] = data.items.map(
    (recipe: Entry<EntrySkeletonType, undefined, string>) => {
      const fields = recipe?.fields as RecipeFields;

      const id: string = recipe.sys.id;
      const title = fields?.title || "";
      const image = (fields?.photo?.fields?.file?.url || "") as string;
      const tags = fields?.tags?.map(
        (tag) => (tag?.fields?.name as string) || ""
      );
      const description = fields?.description || "";
      const chefName = (fields?.chef?.fields?.name || "") as string;

      const recipeData: Recipe = {
        id,
        title,
        image,
        tags,
        description,
        chefName,
      };

      return recipeData;
    }
  );

  return recipes;
};
