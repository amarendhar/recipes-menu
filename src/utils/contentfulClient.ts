import { createClient, CreateClientParams } from "contentful";
import {
  CONTENTFUL_SPACE_ID,
  CONTENTFUL_ACCESS_TOKEN,
  CONTENTFUL_ENVIRONMENT_ID,
} from "globalConstants";

const clientParams: CreateClientParams = {
  space: CONTENTFUL_SPACE_ID,
  accessToken: CONTENTFUL_ACCESS_TOKEN,
};

if (CONTENTFUL_ENVIRONMENT_ID) {
  clientParams.environment = CONTENTFUL_ENVIRONMENT_ID;
}

export const client = createClient(clientParams);
