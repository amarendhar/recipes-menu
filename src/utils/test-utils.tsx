import { Route, Routes, BrowserRouter } from "react-router-dom";
import {
  render as rtlRender,
  renderHook as rtlRenderHook,
} from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { defaultTheme } from "themes";
import { ContentfulProvider } from "api";
import App from "App";

const render = (
  ui: React.ReactElement,
  { route = "/" }: { route?: string | null } = {}
) => {
  window.history.pushState({}, "Test page", route);

  const Wrapper = ({ children }: { children: React.ReactNode }) =>
    route ? (
      <App />
    ) : (
      <ContentfulProvider>
        <ThemeProvider theme={defaultTheme}>
          <BrowserRouter>{children}</BrowserRouter>
        </ThemeProvider>
      </ContentfulProvider>
    );

  return {
    ...rtlRender(ui, {
      wrapper: Wrapper,
    }),
  };
};

const renderHook = (
  hook: any,
  { route = "/" }: { route?: string | null } = {}
) => {
  window.history.pushState({}, "Test page", route);

  const Wrapper = ({ children }: { children: any }) => (
    <ContentfulProvider>
      <BrowserRouter>
        <Routes>
          <Route path="/" element={children} />
          <Route path="/recipe/:id" element={children} />
        </Routes>
      </BrowserRouter>
    </ContentfulProvider>
  );

  return rtlRenderHook(() => hook(), { wrapper: Wrapper });
};

export * from "@testing-library/react";
export { render, renderHook };
