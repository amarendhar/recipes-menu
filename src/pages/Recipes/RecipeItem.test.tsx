import { render, screen } from "utils/test-utils";
import { RecipeItem } from "./RecipeItem";
import { prepareRecipes } from "./utils";
import { RecipeListItem, RecipesData } from "types";
import mockRecipes from "mocks/__fixtures__/mockRecipes.json";

describe("# RecipeItem", () => {
  const recipes = prepareRecipes(mockRecipes as RecipesData);

  const renderComponent = (recipe: RecipeListItem) =>
    render(<RecipeItem recipe={recipe} />, { route: null });

  it("Should render recipe-item with recipe image and title", async () => {
    const recipe = recipes[0];
    const { container } = renderComponent(recipe);

    expect(screen.getByTestId("recipe-item").getAttribute("href")).toEqual(
      `/recipe/${recipe.id}`
    );
    expect(
      (screen.getByTestId("recipe-image") as HTMLImageElement).src
    ).toContain(recipe.image);
    expect(
      (screen.getByTestId("recipe-image") as HTMLImageElement).alt
    ).toEqual(recipe.title);
    expect(screen.getByTestId("recipe-name").textContent).toEqual(recipe.title);

    expect(container).toMatchSnapshot();
  });
});
