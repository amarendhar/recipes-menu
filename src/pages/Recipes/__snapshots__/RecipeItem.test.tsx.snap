// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`# RecipeItem Should render recipe-item with recipe image and title 1`] = `
<div>
  <a
    class=""
    data-testid="recipe-item"
    href="/recipe/4dT8tcb6ukGSIg2YyuGEOm"
  >
    <div
      class="sc-fLljvp drMhkk"
    >
      <div
        class="sc-bBrFuS MUIGI"
      >
        <img
          alt="White Cheddar Grilled Cheese with Cherry Preserves & Basil"
          data-testid="recipe-image"
          src="//images.ctfassets.net/kk2bw5ojx476/61XHcqOBFYAYCGsKugoMYK/0009ec560684b37f7f7abadd66680179/SKU1240_hero-374f8cece3c71f5fcdc939039e00fb96.jpg"
        />
      </div>
      <div
        class="sc-ivTnkv cKcKeu"
      >
        <div
          class="sc-cxaavb gRHQQK"
          data-testid="recipe-name"
        >
          White Cheddar Grilled Cheese with Cherry Preserves & Basil
        </div>
        <div
          class="sc-jqUWZQ kSNpbf"
          data-testid="rating-container"
        >
          <button
            class="sc-kDDqET fgsgTg"
            data-testid="rating-1"
          >
            <svg
              height="24"
              style="color: rgba(0, 0, 0, 0.26);"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </button>
          <button
            class="sc-kDDqET fgsgTg"
            data-testid="rating-2"
          >
            <svg
              height="24"
              style="color: rgba(0, 0, 0, 0.26);"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </button>
          <button
            class="sc-kDDqET fgsgTg"
            data-testid="rating-3"
          >
            <svg
              height="24"
              style="color: rgba(0, 0, 0, 0.26);"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </button>
          <button
            class="sc-kDDqET fgsgTg"
            data-testid="rating-4"
          >
            <svg
              height="24"
              style="color: rgba(0, 0, 0, 0.26);"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </button>
          <button
            class="sc-kDDqET fgsgTg"
            data-testid="rating-5"
          >
            <svg
              height="24"
              style="color: rgba(0, 0, 0, 0.26);"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </button>
          <input
            class="sc-iqcnOu ZBJeF"
            data-testid="rating"
            name="rating"
            type="radio"
            value="0"
          />
        </div>
      </div>
    </div>
  </a>
</div>
`;
